add: marathon.js

amorfati0310 opened the pull request:
프로그래머스 테스트 통과가 안되는데 어느 부분에서 막힌 건지 잘 모르겠어서 issue 공유

참가장 중에 순회를 하면서 완료한 사람에 맞는 사람이 없을 때 idx를 결정하고 
그 idx로 참가자 리스트에서 값을 반환하는데 
어디서 잘 못 된 건지 잘 모르겠어서 도움을 청합니다.

https://programmers.co.kr/learn/courses/30/lessons/42576

amorfati0310 committed:
add: marathon.js

프로그래머스 테스트 통과가 안되는데 어느 부분에서 막힌 건지 잘 모르겠어서 issue 공유
[Ticket: X]

amorfati0310 committed:
fix: some methods forEach로 변경 후 다시 수정

some에서 상태 값을 조작하고 true false로 return 했던 부분 다시 수정 정확한 이유는 추후... 찾아보기

amorfati0310 commented:
프로그래머스 test 에러는 some을 이상하게 써서 나는 에러 같아서 나중에 다시 알아보겠습니다 ;ㅁ;
값을 찾았을 때  js에서 순회해서 바로 답을 return 하려면 일반 for문 으로  return ; 말고 또 없을련지 한 번 더 고민해봐야 겠네요

ahastudio commented on marathon-dali/marathon.js:
```diff
@@ -0,0 +1,13 @@
+  const completionObj = completion.reduce((acc, crr) => {
+    acc[crr] ? acc[crr] += 1 : acc[crr] = 1 
+    return acc;
+  }, {})
```
`reduce`를 쓸 때는 accumulator를 직접 고치지 않는 게 좋습니다.

```javascript
const completionObj = completion.reduce((acc, cur) => ({
    ...acc,
    [cur]: (acc[cur] || 0) + 1,
}), {});
```

amorfati0310 replied:
이렇게 쓰니 훨씬 깔끔하네요 :)  퇴근후 수정해서 다시 pr 날리겠습니당 감사합니다 ㅎㅅㅎ

ahastudio commented on marathon-dali/marathon.js:
```diff
@@ -0,0 +1,13 @@
+  let result = '';
+  participant.forEach( name => {
+    if(completionObj[name]) completionObj[name] -=1 
+    else result = name; 
```
지금은 정답을 초기에 찾아도 끝까지 진행하게 되어 있습니다. `forEach`를 중간에 멈추는 게 불가능하기 때문이죠. `result`가 있으면 `completionObj`에서 탐색하는 일만 줄여도 더 나을 겁니다.

`completionObj`를 조작하지 않고 문제를 해결하고 싶다면 어떻게 해야 할까요?

amorfati0310 replied:
질문
1.result가 있으면 completionObj에서 탐색하는 일만 줄여도 더 나을 겁니다.
```js
if(!result && completionObj[name]) completionObj[name] -=1 
else result = name;
```
이렇게 하면 탐색하는 효과를 줄일 수 있다는 말 맞나요? 
2. completionObj 조작하지 않고 문제 해결 ? 지금은 바로 생각나는게 없는데 고민해서 다시 pr 날려보겠습니다 감사합니다

ahastudio replied:
`else`는 안 쓰는 게 명확합니다.

```javascript
if (result) {
    return;
}

if (completionObj[name] === 0) {
    result = name;
    return;
}

completionObj[name] -= 1;
```

만약 이 코드가 너무 길게 느껴진다면 그건 라인을 억지로 줄이는 게 아니라 다른 로직을 써야 하는 상황인 겁니다.

amorfati0310 replied:
감사합니다 :)

amorfati0310 committed:
refacotr: review 반영
1. reduce 값만 가지고 오고 직접 쓰지 않게
2. result를 구할 때도 같은 방식으로

다만 삼항연산자 중첩인 점이 아쉬움 풀어 쓰는게 더 가독성 좋은 코드일까 고민
